File: src/SuperwallOptions.ts
/**
 * @category Enums
 * @since 0.0.15
 * Defines the log levels for the SDK.
 */
export enum LogLevel {
  Debug = "debug",
  Info = "info",
  Warn = "warn",
  Error = "error",
  None = "none",
}

/**
 * @category Enums
 * @since 0.0.15
 * Defines the scopes for logging within the SDK.
 */
export enum LogScope {
  LocalizationManager = "localizationManager",
  BounceButton = "bounceButton",
  CoreData = "coreData",
  ConfigManager = "configManager",
  IdentityManager = "identityManager",
  DebugManager = "debugManager",
  DebugViewController = "debugViewController",
  LocalizationViewController = "localizationViewController",
  GameControllerManager = "gameControllerManager",
  Device = "device",
  Network = "network",
  PaywallEvents = "paywallEvents",
  ProductsManager = "productsManager",
  StoreKitManager = "storeKitManager",
  Placements = "placements",
  Receipts = "receipts",
  SuperwallCore = "superwallCore",
  PaywallPresentation = "paywallPresentation",
  PaywallTransactions = "paywallTransactions",
  PaywallViewController = "paywallViewController",
  Cache = "cache",
  All = "all",
}

/**
 * @category Enums
 * @since 0.0.15
 * Defines the network environment for Superwall.
 */
export enum NetworkEnvironment {
  Release = "release",
  ReleaseCandidate = "releaseCandidate",
  Developer = "developer",
}

/**
 * @category Enums
 * @since 0.0.15
 * Defines the different types of views that can appear behind Apple's payment sheet during a transaction.
 */
export enum TransactionBackgroundView {
  spinner = "spinner",
  none = "none",
}

/**
 * @category Models
 * @since 0.0.15
 * Defines the messaging of the alert presented to the user when restoring a transaction fails.
 */
export interface RestoreFailed {
  title: string
  message: string
  closeButtonTitle: string
}

/**
 * @category Models
 * @since 0.0.15
 * Options for configuring logging behavior.
 */
export interface LoggingOptions {
  level: LogLevel
  scopes: LogScope[]
}

/**
 * @category Models
 * @since 0.0.15
 * Options for configuring the appearance and behavior of paywalls.
 */
export interface PaywallOptions {
  isHapticFeedbackEnabled: boolean
  restoreFailed: RestoreFailed
  shouldShowPurchaseFailureAlert: boolean
  shouldPreload: boolean
  automaticallyDismiss: boolean
  transactionBackgroundView: TransactionBackgroundView
}

/**
 * @category Models
 * @since 0.0.15
 * Options for configuring the Superwall SDK.
 */
export interface SuperwallOptions {
  paywalls: PaywallOptions
  networkEnvironment: NetworkEnvironment
  isExternalDataCollectionEnabled: boolean
  localeIdentifier?: string
  isGameControllerEnabled: boolean
  logging: LoggingOptions
  collectAdServicesAttribution: boolean
  passIdentifiersToPlayStore: boolean
  storeKitVersion?: "STOREKIT1" | "STOREKIT2"
  enableExperimentalDeviceVariables: boolean
  manualPurchaseManagement: boolean
}

/**
 * @category Models
 * @since 0.0.15
 * Default options for the Superwall SDK.
 */
export const DefaultSuperwallOptions: SuperwallOptions = {
  paywalls: {
    isHapticFeedbackEnabled: true,
    restoreFailed: {
      title: "No Subscription Found",
      message: "We couldn't find an active subscription for your account.",
      closeButtonTitle: "Okay",
    },
    shouldShowPurchaseFailureAlert: true,
    shouldPreload: false,
    automaticallyDismiss: true,
    transactionBackgroundView: TransactionBackgroundView.spinner,
  },
  networkEnvironment: NetworkEnvironment.Release,
  isExternalDataCollectionEnabled: true,
  isGameControllerEnabled: false,
  logging: {
    level: LogLevel.Info,
    scopes: [LogScope.All],
  },
  collectAdServicesAttribution: false,
  passIdentifiersToPlayStore: false,
  enableExperimentalDeviceVariables: false,
  manualPurchaseManagement: false,
}

File: android/src/main/java/expo/modules/superwallexpo/json/SuperwallOptions.kt
package expo.modules.superwallexpo.json


import com.superwall.sdk.config.options.PaywallOptions
import com.superwall.sdk.config.options.SuperwallOptions
import com.superwall.sdk.logger.LogLevel
import com.superwall.sdk.logger.LogScope
import java.util.EnumSet

fun superwallOptionsFromJson(json: Map<String, Any?>): SuperwallOptions {
  val options = SuperwallOptions()
  options.localeIdentifier = json["localeIdentifier"] as String?
  options.isExternalDataCollectionEnabled = (json["isExternalDataCollectionEnabled"] as Boolean?)?:true
  options.isGameControllerEnabled = (json["isGameControllerEnabled"] as Boolean?)?:false
  options.passIdentifiersToPlayStore = (json["passIdentifiersToPlayStore"] as Boolean?)?:false
  options.enableExperimentalDeviceVariables = (json["enableExperimentalDeviceVariables"] as Boolean?)?:false
  options.shouldObservePurchases = (json["shouldObservePurchases"] as Boolean?)?:false
  options.useMockReviews = (json["useMockReviews"] as Boolean?)?:false

      val networkEnvironment = when (json["networkEnvironment"] as String?) {
        "release" -> SuperwallOptions.NetworkEnvironment.Release()
        "releaseCandidate" -> SuperwallOptions.NetworkEnvironment.ReleaseCandidate()
        "developer" -> SuperwallOptions.NetworkEnvironment.Developer()
        else -> options.networkEnvironment
      }
      options.networkEnvironment = networkEnvironment

      // Logging
      val loggingMap = json["logging"] as Map<String, Any>?
      val scopesArray = loggingMap?.get("scopes") as List<String>?
      val scopes = scopesArray?.map {
        try {
          LogScope.valueOf(it)
        } catch (e: IllegalArgumentException) {
          null
        }
      }?.filterNotNull() ?: emptyList()

      val levelStr = loggingMap?.get("level") as String?
      val level = LogLevel.values().find { it.toString().equals(levelStr, ignoreCase = true) }
        ?: LogLevel.warn

      val logging = SuperwallOptions.Logging()
      if(scopes.isNotEmpty()) {
        logging.scopes = EnumSet.copyOf(scopes)
      }
      logging.level = level
      options.logging = logging

      // PaywallOptions
      val paywallsMap = json["paywalls"] as Map<String, Any>?
      val isHapticFeedbackEnabled = paywallsMap?.get("isHapticFeedbackEnabled") as Boolean?
      val shouldShowPurchaseFailureAlert = paywallsMap?.get("shouldShowPurchaseFailureAlert") as Boolean?
      val shouldPreload = paywallsMap?.get("shouldPreload") as Boolean?
      val automaticallyDismiss = paywallsMap?.get("automaticallyDismiss") as Boolean?

      val paywalls = PaywallOptions()
      paywalls.isHapticFeedbackEnabled = isHapticFeedbackEnabled ?: paywalls.isHapticFeedbackEnabled
      paywalls.shouldShowPurchaseFailureAlert = shouldShowPurchaseFailureAlert ?: paywalls.shouldShowPurchaseFailureAlert
      paywalls.shouldPreload = shouldPreload ?: paywalls.shouldPreload
      paywalls.automaticallyDismiss = automaticallyDismiss ?: paywalls.automaticallyDismiss
      // Note: shouldShowWebPurchaseConfirmationAlert is iOS-only

      val restoreFailedMap = paywallsMap?.get("restoreFailed") as Map<String, Any>?
      val restoreFailed = PaywallOptions.RestoreFailed().apply {
        restoreFailedMap?.let {
          title = it["title"] as String? ?: title // Keep default if not found
          message = it["message"] as String? ?: message // Keep default if not found
          closeButtonTitle = it["closeButtonTitle"] as String? ?: closeButtonTitle // Keep default if not found
        }
      }
      paywalls.restoreFailed = restoreFailed

      val transactionBackgroundViewStr = paywallsMap?.get("transactionBackgroundView") as String?
      paywalls.transactionBackgroundView = when (transactionBackgroundViewStr) {
        "spinner" -> PaywallOptions.TransactionBackgroundView.SPINNER
        else -> null
      }

  options.paywalls = paywalls

  return options
}
